scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -193,7 +192,24 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
     transform: PTransform[_ >: PCollection[T], PCollection[U]]
   ): SCollection[U] = {
     val coder = CoderMaterializer.beam(context, Coder[U])
-    ensureSerializable(coder).fold(throw _, pApply(name, transform).setCoder)
+    val xform: PTransform[_ >: PCollection[T], PCollection[U]] = {
+      if (!context.isTest) {
+        transform
+      } else {
+        val id = context.testId.get
+        val optXForm = TestDataManager
+          .getTransform(id)
+          .flatMap { t =>
+            t.apply[T, U](
+              MockTransform[T, U](name),
+              internal.getCoder,
+              coder
+            )
+          }
+        optXForm.getOrElse(transform)
+      }
+    }
+    ensureSerializable(coder).fold(throw _, pApply(name, xform).setCoder)
```

Comment:
Beam already has an API to override transforms in a Pipeline. see [Pipeline.html#replaceAll](https://beam.apache.org/releases/javadoc/2.0.0/org/apache/beam/sdk/Pipeline.html#replaceAll-java.util.List-)